### ISSUE TITLE
Uninitialized lateinit property

### ISSUE THREAD

**Issue by santiJourni** [Adyen/adyen-android#480]
**Describe the bug**
The `lateinit` property `BaseComponentDialogFragment#component` is not initialized when used by `GenericComponentDialogFragment`.

**To Reproduce**
I haven't been able to reproduce the crash. I have tried forcing screen restorations using the "Discard Activities" toggle, but haven't been able to successfully reproduce it.

**Expected behavior**
The property should be initialized before use

**Smartphone (please complete the following information):**
 - Device: HUAWEI P smart 2019
 - OS: Android 10
 - Browser: N.A. Don't have access to this info in our crash logs
 - Version: (Adyen version): 4.1.1

**Additional context**
StackTrace:
```
Fatal Exception: kotlin.UninitializedPropertyAccessException lateinit property component has not been initialized
com.adyen.checkout.dropin.ui.base.BaseComponentDialogFragment.getComponent (BaseComponentDialogFragment.java:50)
com.adyen.checkout.dropin.ui.component.GenericComponentDialogFragment.onViewCreated (GenericComponentDialogFragment.java:70)
androidx.fragment.app.Fragment.performViewCreated (Fragment.java:2987)
androidx.fragment.app.FragmentStateManager.createView (FragmentStateManager.java:546)
androidx.fragment.app.FragmentStateManager.moveToExpectedState (FragmentStateManager.java:282)
androidx.fragment.app.FragmentManager.executeOpsTogether (FragmentManager.java:2189)
androidx.fragment.app.FragmentManager.removeRedundantOperationsAndExecute (FragmentManager.java:2100)
androidx.fragment.app.FragmentManager.execPendingActions (FragmentManager.java:2002)
androidx.fragment.app.FragmentManager$5.run (FragmentManager.java:524)
android.os.Handler.handleCallback (Handler.java:888)
android.os.Handler.dispatchMessage (Handler.java:100)
android.os.Looper.loop (Looper.java:213)
android.app.ActivityThread.main (ActivityThread.java:8178)
java.lang.reflect.Method.invoke (Method.java)
com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run (RuntimeInit.java:513)
com.android.internal.os.ZygoteInit.main (ZygoteInit.java:1101) 
```

**Comment by jreij** [comment]
Hi, thanks for notifying us about this, I need a few more details though. How does "Discard Activities" work exactly? Do you put the app in the background then come back to it while Discard Activities is on? And if so, which screen are you on right before you put the app in the background?

I'm asking because we test using the "Don't keep activities" toggle in the developer options but we can't reproduce this issue.

**Comment by santiJourni** [comment]
What I did to try to reproduce the issue is to put the app in the background with the "Don't keep activities" toggle on while in the Adyen Drop In and also during various steps of the payment process eg. while in the webview to pay with paypal, but as I said I wasn't able to reproduce the issue.

Unfortunately, one of our users gets this error consistently. Our logs report that they got the error 7 times in a row despite starting the whole payment process again.

**Comment by santiJourni** [comment]
This issue is becoming more common. Over the weekend 6 more users had the same problem. The issue occurs at different moments in the process. For some users it happens in the original DropIn and for others it happens when required to take additional action. Regardless the crash is the same.

I unfortunately still haven't been able to reproduce it, but all users that experience the issue are from Belgium.

Also, some users were able to complete the payment after some retries while others couldn't despite retrying 7 times.

**Comment by jreij** [comment]
Thanks for the feedback! We will check this again and see what we can find out.

**Comment by santiJourni** [comment]
It is definitely a Bancontact issue. As soon as I click on the Bancontact payment method the drop in crashes

**Comment by jreij** [comment]
Can you check the logs prior to the crash? Is there any error log or something with tag: `BaseComponentDialogFragment` ?

**Comment by santiJourni** [comment]
I haven't seen any error log with `BaseComponentDialogFragment` as the tag, but there is one error:

```
2021-09-20 10:08:36.604 3144-11835/? E/bfrk: Exception occurred while getting SettingsStore instance.
    java.util.concurrent.ExecutionException: com.google.android.gms.auth.UserRecoverableAuthException: BadAuthentication
        at celt.s(:com.google.android.gms@213314021@21.33.14 (040408-395723304):3)
        at celt.get(:com.google.android.gms@213314021@21.33.14 (040408-395723304):13)
        at bfrk.c(:com.google.android.gms@213314021@21.33.14 (040408-395723304):3)
        at bfrk.a(:com.google.android.gms@213314021@21.33.14 (040408-395723304):1)
        at bfnd.c(:com.google.android.gms@213314021@21.33.14 (040408-395723304):3)
        at bfnd.f(:com.google.android.gms@213314021@21.33.14 (040408-395723304):16)
        at afcb.run(:com.google.android.gms@213314021@21.33.14 (040408-395723304):17)
        at bywp.run(:com.google.android.gms@213314021@21.33.14 (040408-395723304):2)
        at wuw.c(:com.google.android.gms@213314021@21.33.14 (040408-395723304):6)
        at wuw.run(:com.google.android.gms@213314021@21.33.14 (040408-395723304):7)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
        at xav.run(:com.google.android.gms@213314021@21.33.14 (040408-395723304):0)
        at java.lang.Thread.run(Thread.java:784)
     Caused by: com.google.android.gms.auth.UserRecoverableAuthException: BadAuthentication
        at hyp.p(:com.google.android.gms@213314021@21.33.14 (040408-395723304):3)
        at hyi.a(:com.google.android.gms@213314021@21.33.14 (040408-395723304):6)
        at hyp.r(:com.google.android.gms@213314021@21.33.14 (040408-395723304):4)
        at hyp.q(:com.google.android.gms@213314021@21.33.14 (040408-395723304):15)
        at hyp.v(:com.google.android.gms@213314021@21.33.14 (040408-395723304):0)
        at hyp.w(:com.google.android.gms@213314021@21.33.14 (040408-395723304):0)
        at hyp.t(:com.google.android.gms@213314021@21.33.14 (040408-395723304):0)
        at bfrx.call(:com.google.android.gms@213314021@21.33.14 (040408-395723304):3)
        at cepn.a(:com.google.android.gms@213314021@21.33.14 (040408-395723304):0)
        at ceoo.run(:com.google.android.gms@213314021@21.33.14 (040408-395723304):3)
        at cepo.run(:com.google.android.gms@213314021@21.33.14 (040408-395723304):0)
        at ceov.execute(:com.google.android.gms@213314021@21.33.14 (040408-395723304):5)
        at java.util.concurrent.AbstractExecutorService.submit(AbstractExecutorService.java:140)
        at celu.e(:com.google.android.gms@213314021@21.33.14 (040408-395723304):0)
        at bfry.a(:com.google.android.gms@213314021@21.33.14 (040408-395723304):3)
        at bfnd.c(:com.google.android.gms@213314021@21.33.14 (040408-395723304):2)
        at bfnd.f(:com.google.android.gms@213314021@21.33.14 (040408-395723304):16) 
        at afcb.run(:com.google.android.gms@213314021@21.33.14 (040408-395723304):17) 
        at bywp.run(:com.google.android.gms@213314021@21.33.14 (040408-395723304):2) 
        at wuw.c(:com.google.android.gms@213314021@21.33.14 (040408-395723304):6) 
        at wuw.run(:com.google.android.gms@213314021@21.33.14 (040408-395723304):7) 
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162) 
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636) 
        at xav.run(:com.google.android.gms@213314021@21.33.14 (040408-395723304):0) 
        at java.lang.Thread.run(Thread.java:784) 
```

**Comment by jreij** [comment]
Hmm I don't think this is related. Can you check what is the payment method type for BCMC inside the `PaymentMethodsApiResponse` that you pass for `DropIn.startPayment`?

Should be something like this:
```json
{
      "name": "Bancontact card",
      "type": "bcmc"
    }
```

**Comment by santiJourni** [comment]
What should I check in the response? 

The response contains exactly that: 

```
{
    "name": "Bancontact card",
    "type": "bcmc"
}
```

and the credit card payment method also includes:

```
{
    "brands": [
        "bcmc",
        ....
     ],
     ....
}
```

**Comment by jreij** [comment]
Yeah that's what I meant, nothing seems wrong there 🤔  I have a few more questions if you don't mind:

> As soon as I click on the Bancontact payment method the drop in crashes

Can you reproduce this every time, even with "Don't keep activities" switched off? And with which device are you testing now?

**Comment by santiJourni** [comment]
It happens every time even with "Don't keep activities" switched off. 

The testing device is a:

HUAWEI Mate 10 lite
model RNE-L21 
running android version 8.0.0

**Comment by jreij** [comment]
Are you sure you cannot see any error logs prior to the crash? Specifically looking for something like: `Unable to find component configuration for paymentMethod - bcmc`

**Comment by santiJourni** [comment]
No, I don't see any error log like that. I have checked all error levels starting from Debug and can't see any thing related to Adyen crashing or reporting an issue

**Comment by jreij** [comment]
Ah yes, you need to call `com.adyen.checkout.core.log.Logger.setLogcatLevel` in your application class.

**Comment by jreij** [comment]
In all cases, I think I found the issue. To be sure, try this workaround:
* Inside the method where you create `DropInConfiguration.Builder`, create a `BcmcConfiguration`. And example could be:
```kotlin
val bcmcConfiguration = BcmcConfiguration.Builder(shopperLocale, environment, clientKey)
            .build()
```
* Make sure that `shopperLocale`, `environment` and `clientKey` match the values used in `DropInConfiguration.Builder`
* Call `addBcmcConfiguration(bcmcConfiguration)` on your `DropInConfiguration.Builder`

If this indeed does fix the issue then feel free to make a release with this workaround and we will fix this ourselves in the next release so you can remove the workaround afterwards.

**Comment by santiJourni** [comment]
This work around seems to fix the issue. Thanks for the quick responses and fix

**Comment by jreij** [comment]
Sure, glad I could help 🙂

**Comment by jreij** [comment]
@santiJourni we just released 4.2.0 which includes the fix for this issue, it should be live in maven central a bit. If you upgrade to this version you can remove the `BcmcConfiguration` workaround 😄
